components/ai_chat/core/browser/tools/memory_storage_tool.cc

```diff
@@ -0,0 +1,118 @@
+// Copyright (c) 2025 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#include "brave/components/ai_chat/core/browser/tools/memory_storage_tool.h"
+
+#include <utility>
+
+#include "base/json/json_reader.h"
+#include "base/strings/strcat.h"
+#include "base/strings/string_number_conversions.h"
+#include "base/values.h"
+#include "brave/components/ai_chat/core/browser/tools/tool_input_properties.h"
+#include "brave/components/ai_chat/core/common/mojom/ai_chat.mojom.h"
+#include "brave/components/ai_chat/core/common/prefs.h"
+#include "components/prefs/pref_service.h"
+
+namespace ai_chat {
+
+MemoryStorageTool::MemoryStorageTool(PrefService* pref_service)
+    : pref_service_(pref_service) {
+  CHECK(pref_service);
+}
+
+MemoryStorageTool::~MemoryStorageTool() = default;
+
+std::string_view MemoryStorageTool::Name() const {
+  return mojom::kMemoryStorageToolName;
+}
+
+std::string_view MemoryStorageTool::Description() const {
+  return "Store user information ONLY when user explicitly requests to "
+         "remember something "
+         "with phrases like 'Remember that I...', 'Please note that I...', "
+         "'Store that I...', "
+         "or similar direct memory commands. Do NOT use for casual mentions, "
+         "examples, "
+         "questions, or context. Returns empty string on success, error "
+         "message "
+         "on failure. Call this tool at most once per turn.";
+}
```

Comment:
Is there a reason we're forcing the AI to only suggest a memory if the user explicitly asks it to save something specific? It seems like it might not get used very often. Was that a requirement? Or did you perhaps notice it being used too often with a more permissive tool description? If not, we could try having it only save facts about the user it deems important for future conversations, to aid the user not having to repeat themselves. Otherwise, I don't think this tool will get used very much.

Reply:
It's quite often and annoying without this instruction, so I added it back then. I think I did try to include instruction like let them decide what's worth to be saved, but didn't seem to work well back then. So figure I could start this first and get all the pieces in then tweaking again.